README.md
# react-native-stylesheet-xg
extension stylesheet for cross platforms and responsive

## Example
Check [index.js](https://github.com/xgfe/react-native-stylesheet-xg/blob/master/example/index.android.js) in the Example folder.

### origin
```javascript
{
  test1: {
    width: 0,
    height: 1,
    borderWidth: 100,
    $borderRadius: 100,
  },
  test: {
    width: 100,
    color: 'red',
    android: {
      width: 0,
      height: 1,
      width: 200,
      color: 'blue'
    },
    ios: {
      width: 300,
      color: 'green'
    }
  }
}
```

### android (width: 360, base: 320)
```javascript
{
  test1: {
    width: 0,
    height: 1,
    borderWidth: 113,
    $borderRadius: 100,
  },
  test: {
    width: 225,
    color: 'blue',
    height: 1
  }
}
```

### ios (width: 320, base: 320)
```javascript
{
  test1: {
    width: 0,
    height: 1,
    borderWidth: 100,
    $borderRadius: 100,
  },
  test: {
    width: 300,
    color: 'green'
  }
}
```

## Usage

```bash
npm install react-native-stylesheet-xg --save
```

```javascript
import StyleSheet from 'react-native-stylesheet-xg';

// before use, you may need set the base width with StyleSheet.setBase(width:number), which default to 320
StyleSheet.setBase(360); // this is optional, if your target screen is 320

// use react-native-stylesheet-xg replace react-native StyleSheet Module
StyleSheet.create({
  test: {
    width: 100
  }
});

// or you can use StyleSheet.r(num:number) to get the target responsive num
StyleSheet.r(100); // 113 (width: 360, base: 320)

```

| Method  | Params  | Description |
| :------------ |:---------------:| :---------------:|
| setBase | width:number | set the target width according to the UI |
| setMinWidth | width:number | set the responsive minWidth. If the device width is less than minWidth, will use minWidth as the device width! |
| setMaxWidth | width:number | set the responsive maxWidth. If the device width is greater than minWidth, will use maxWidth as the device width! |
| create | style:object | replace native StyleSheet creating stylesheet  |
| r | num:number | the inner function to get the responsive size according to the device with and base |
| switchR | isRon:boolean | set the flag to switch the responsive size function, which return the prev value |


CHANGELOG.md
<a name="1.1.0"></a>
# [1.1.0](https://github.com/xgfe/react-native-stylesheet-xg/compare/v1.0.3...v1.1.0) (2016-10-26)


### Features

* **StyleSheet:** add `minWidth` and `maxWidth` support ([3be5bc7](https://github.com/xgfe/react-native-stylesheet-xg/commit/3be5bc7))



<a name="1.0.3"></a>
## [1.0.3](https://github.com/xgfe/react-native-stylesheet-xg/compare/v1.0.1...v1.0.3) (2016-05-02)


### Features

* **stylesheet:** add support to turn on/off the responsive size function ([8f97a8e](https://github.com/xgfe/react-native-stylesheet-xg/commit/8f97a8e))



<a name="1.0.1"></a>
## [1.0.1](https://github.com/xgfe/react-native-stylesheet-xg/compare/v1.0.0...v1.0.1) (2016-05-02)


### Bug Fixes

* **stylesheet:** fix the negative number revise logic ([e2e28b8](https://github.com/xgfe/react-native-stylesheet-xg/commit/e2e28b8))



<a name="1.0.0"></a>
# 1.0.0 (2016-05-01)





index.js
'use strict';
import React, {
  StyleSheet,
  Dimensions,
  Platform
} from 'react-native';

const {width} = Dimensions.get('window');
const DEFAULT_WIDTH = 320;

const CStyleSheet = {
  minWidth: undefined,
  maxWidth: undefined,
  baseWidth: DEFAULT_WIDTH,
  ratio: width / DEFAULT_WIDTH,
  isROn: true,
  setBase: function (baseWidth = DEFAULT_WIDTH) {
    this.baseWidth = baseWidth;
    this.updateRatio();
  },
  setMinWidth: function (minWidth) {
    this.minWidth = minWidth;
    this.updateRatio();
  },
  setMaxWidth: function (maxWidth) {
    this.maxWidth = maxWidth;
    this.updateRatio();
  },
  updateRatio: function () {
    if (width < this.minWidth) {
      this.ratio = this.minWidth / this.baseWidth;
    } else if (width > this.maxWidth) {
      this.ratio = this.maxWidth / this.baseWidth;
    } else {
      this.ratio = width / this.baseWidth;
    }
  },
  create: function (styleObj) {
    return StyleSheet.create(this.revise(styleObj));
  },

  revise: function(styleObj) {
    // revise, expect: flex, opacity
    for (var styleKey in styleObj) {
      // Freeze keys
      var _keys = Object.keys(styleObj[styleKey]);
      for (var i in _keys) {
        var key = _keys[i];
        // cross platform
        if (key === 'ios' || key === 'android') {
          if (key === Platform.OS) {
            for (var platformKey in styleObj[styleKey][key]) {

              if (typeof styleObj[styleKey][key][platformKey] === 'number' && platformKey !== 'flex' && platformKey !== 'opacity') {
                // ignore rules with '$' start
                if (platformKey.charAt(0) === '$') {
                  styleObj[styleKey][platformKey.substr(1)] = styleObj[styleKey][key][platformKey];
                  continue;
                }
                styleObj[styleKey][platformKey] = this.r(styleObj[styleKey][key][platformKey]);
              } else {
                // directly assign to the parent
                styleObj[styleKey][platformKey] = styleObj[styleKey][key][platformKey];
              }
            }

          }

          // delete hack style key
          delete styleObj[styleKey][key];
          continue;
        }

        if (typeof styleObj[styleKey][key] === 'number' && key !== 'flex' && key !== 'opacity') {
          // ignore rules with '$' start
          if (key.charAt(0) === '$') {
            styleObj[styleKey][key.substr(1)] = styleObj[styleKey][key];
            delete styleObj[styleKey][key];
            continue;
          } else {
            styleObj[styleKey][key] = this.r(styleObj[styleKey][key]);
          }
        }
      }
    }

    return styleObj;
  },

  switchR: function(isROn) {
    var prevStatus = this.isROn;

    this.isROn = !!isROn;

    return prevStatus;
  },

  r: function(num) {
    if (this.isROn && Math.abs(num) > 1) {
      return Math.round(num * this.ratio);
    }

    return num;
  }
};

export default CStyleSheet;


example/index.android.js
/**
 * Sample React Native App
 * https://github.com/facebook/react-native
 */

import React, {
  AppRegistry,
  Component,
  Text,
  View,
  ScrollView,
} from 'react-native';
import StyleSheet from 'react-native-stylesheet-xg';
import JsonBeautify from 'json-beautify';

let testStyle = {
  test1: {
    width: 0,
    height: 1,
    borderWidth: 100,
    $borderRadius: 100,
  },
  test: {
    width: 100,
    color: 'red',
    android: {
      width: 0,
      height: 1,
      width: 200,
      color: 'blue'
    },
    ios: {
      width: 300,
      color: 'green'
    }
  }
}

class example extends Component {
  render() {
    return (
      <View style={style.container}>
        <Text style={style.title}>react-native-stylesheet-xg</Text>
        <View style={style.body}>
          <View style={style.textCon}>
            <ScrollView>
              <Text style={{color: '#333'}}>{JsonBeautify(testStyle, null, 3, 10)}</Text>
            </ScrollView>
          </View>

          <View style={style.textCon}>
            <ScrollView>
              <Text style={{color: '#333'}}>{JsonBeautify(StyleSheet.revise(testStyle), null, 3, 10)}</Text>
            </ScrollView>
          </View>
        </View>
      </View>
    );
  }
};

const style = StyleSheet.create({
  container: {
    flexDirection: 'column',
    flex: 1,
    ios: {
      marginTop: 20
    }
  },
  title: {
    $fontSize: 20,
    color: '#333',
    textAlign: 'center'
  },
  body: {
    flex: 1,
    flexDirection: 'row'
  },
  textCon: {
    flex: 1,
    margin: 5,
    borderWidth: 1,
    borderColor: '#eee'
  }
});

AppRegistry.registerComponent('example', () => example);
